streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -757,6 +752,23 @@ private static HostInfo parseHostInfo(final String endPoint) {
         return new HostInfo(host, port);
     }
 
+    private boolean setRunningFromCreated() {
+        synchronized (stateLock) {
+            if (state != State.CREATED) {
+                throw new IllegalStateException("Stream-client " + clientId + ": Unexpected state transition from " + state + " to " + State.RUNNING);
+            }
+            state = State.RUNNING;
+            stateLock.notifyAll();
+        }
+
+        // we need to call the user customized state listener outside the state lock to avoid potential deadlocks
```

Comment:
Does this mean that the state listener might be invoked out of order?
Eg:
1. RUNNING -> REBALANCING
2. CREATED -> RUNNING

Maybe this is already accounted for, and it's why we pass both the old and new state?

Reply:
Yes it is possible that user's callback maybe triggered out of ordering (in practice may be not very likely), this is not ideal but at that time we did not come up with a better idea how to enforce that we always have instance-level state to go CREATED -> RUNNING -> REBALANCING -> RUNNING at the starting up phase.

All ears if you have a better idea :)

Reply:
Why could we not fix this? The underlying issue seems to be the state transitions of `StreamThread` -- it allows to go from `CREATED -> RUNNING` -- if we change this, and we can only go to `CREATED -> PARTITION REVOKED` we should be able to tackle this issue? (Maybe a different PR to do this though.)